### PR TITLE
`Development`: Extend documentation for server and client guidelines to include DTOs and WebSocket subscriptions

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context

Many REST responses and WebSocket messages that we send from the server to the client contain a lot of unnecessary information, which increases payload sizes and load on the corresponding systems. As a first step to prevent this in the future (and refactor existing REST endpoints and WebSocket topics) I extended the documentation and inform developers about best practices.

### Description

I added a section about DTOs within the server guidelines and a section for web socket subscriptions within the client guidelines (see screenshots).

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2

### Screenshots

Section about WebSocket subscriptions within the client guidelines:
client

Section about DTOs within the server guidelines:
server p1
server p2

### CODE REVIEW

**Review by terlan98 — COMMENTED** [review]
The changes look good overall, but I have 2 suggestions for improvement 🙂

File: docs/dev/guidelines/client.rst [hunk 1]
```diff
@@ -384,3 +384,14 @@ Do not use the following for this purpose if it can be avoided:
             <!-- Do not do this -->
         </div>
     </div>
+
+17. WebSocket Subscriptions
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply docs/dev/guidelines/client.rst:388]: I only use WebSocket now.

**Review by terlan98 — DISMISSED** [review]
Thanks for addressing my comments. Looks good 👍

File: docs/dev/guidelines/client.rst [hunk 2]
```diff
@@ -384,3 +384,14 @@ Do not use the following for this purpose if it can be avoided:
...
+===========================
+
+The client must not subscribe to more than 20 WebSocket topics simultaneously, regardless of the amount of exercises, lectures, courses, etc. there are for one particular user.
+
+Best Practices:
+
+1. Dynamic Subscription Handling: Subscribe to topics on an as-needed basis. Unsubscribe from topics that are no longer needed to keep the number of active subscriptions within the recommended limit.
+2. Efficient Topic Aggregation: Where possible, use topic aggregation techniques to consolidate related data streams into a single subscription. Consequently, don't create a new topic if there already exists a suitable topic that can be reused.
```
Comment (laadvo) [inline docs/dev/guidelines/client.rst:396]:
```suggestion
2. Efficient Topic Aggregation: Use topic aggregation techniques to consolidate related data streams into a single subscription wherever possible. Consequently, don't create a new topic if an existing one can be reused.
```

File: docs/dev/guidelines/server.rst [hunk 3]
```diff
@@ -137,14 +138,143 @@ Additional notes on the controller methods:
     * Bad Request - the request was wrong.
     * Not Found - can't find the requested data or it should be not accessible yet.
 
-12. Dependency injection
+.. _server-guideline-dto-usage:
+
+12. Use DTOs for data transfer
+==============================
+
+Use data transfer objects (DTOs) to send data from the server to the client, i.e. responses of RestControllers and WebSocket messages. A DTO ...
```
Comment (laadvo) [inline docs/dev/guidelines/server.rst:146]: I'm not sure what the point of this `...` is

File: docs/dev/guidelines/server.rst [hunk 4]
```diff
@@ -137,14 +138,143 @@ Additional notes on the controller methods:
...
+
+Definition and Characteristics of DTOs:
+
+1. Java Records: DTOs are implemented as Java records. Records do not support inheritance, so duplication for DTOs is ok.
+2. No Entity Objects: DTOs must not contain entity objects, but only primitive types (or the corresponding wrapper classes), enums or other DTOs.
+3. Minimal Data: DTOs should only include the minimum amount of data required by the client application. Avoid adding any unnecessary or redundant information.
+4. Single Responsibility: Keep DTOs focused on specific tasks and data subsets to maintain a clear and concise data representation. Avoid using a single DTO for multiple payloads unless the data transferred is exactly the same. Create separate records for new or updated payloads.
+5. Avoid Adding Methods: Refrain from adding methods to DTOs. They should serve only as simple data containers without any business logic.
```
Comment (laadvo) [inline docs/dev/guidelines/server.rst:154]:
```suggestion
5. Avoid Adding Methods: Refrain from adding methods to DTOs. They should only serve as simple data containers without any business logic.
```

**Review by laadvo — COMMENTED** [review]
The new docs look good, I just have two suggestions you could implement and something I didn't  completely understand

File: docs/dev/guidelines/server.rst (hunk 3 above)
Comment (lennart-keller) [reply docs/dev/guidelines/server.rst:146]: Originally, the layout was a bit different. I forgot do delete "A DTO ...". I deleted it now.

**Review by MaximilianAnzinger — APPROVED** [review]
LGTM

**Review by laadvo — APPROVED** [review]
Thanks for implementing my suggestions! Looks good to me :)

**Review by terlan98 — APPROVED** [review]
Recent changes look good. Reapproved 👍

**Pull request merged**